### ISSUE TITLE
[Feature Request] Control the progress of Iterator.get_next()

### ISSUE THREAD

**Issue by Hoeze** [tensorflow/tensorflow#19143]
I just noticed that Iterator.get_next() did not change my batch samples in a tf.while_loop.
However, I'd like to run the training operation in a tf.while_loop, so I need to control the progress of Iterator.get_next() directly.

**My Feature Request:**
It would be cool, if Iterator would provide an operation which can be called to advance its state.

------------------------

### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 18.04
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: 1.8.0
- **Python version**:  3.6.5
- **Bazel version (if compiling from source)**: None
- **GCC/Compiler version (if compiling from source)**: None
- **CUDA/cuDNN version**: None
- **GPU model and memory**: None
- **Exact command to reproduce**: None

**Comment by mrry** [comment]
> I just noticed that Iterator.get_next() did not change my batch samples in a tf.while_loop.

Can you explain what you actually tried? If the `get_next()` op is created inside the while loop body, it should advance in each iteration.

**Comment by Hoeze** [comment]
I opened a question on Stackoverflow to this question:
https://stackoverflow.com/questions/50237486/tf-data-iterator-get-next-how-to-advance-in-tf-while-loop
My problem is that I have to assign the new values to my model.

I also tried the following:
```
a = tf.Variable(iterator.get_next())
while (...): tf.assign(a, iterator.get_next())
```
However, this did not work as well. I still did not get other values in operations using `a`.

**Comment by mrry** [comment]
This question is better handled on Stack Overflow. Unfortunately neither we nor the folks on Stack Overflow will be able to help without a runnable example that exhibits the problem you're facing. The [answer](https://stackoverflow.com/a/50237643/3574081) to your Stack Overflow question explains that the `iterator.get_next()` call must be performed inside the `tf.while_loop()` body, and the same principle should apply to your model code if you wrap it in a `tf.while_loop()`.

**Comment by Hoeze** [comment]
The question on Stackoverflow should be seen as an workaround.

What I'd like to request as a new feature would be the following:
```
a = iterator.get_next()
loss = some_fancy_model(a)
tf.while_loop(...):
    with tf.control_dependency([iterator.step_op]):
        minimize(...)
```

In my opinion this is a much cleaner option as it removes the necessity to define the whole model inside a while_loop.

**Comment by mrry** [comment]
I'd recommend checking out [eager execution mode](https://www.tensorflow.org/programmers_guide/eager), which lets you write code similar to that using a regular Python `while` (or `for`) loop.

**Comment by Hoeze** [comment]
I'm sorry for the wrong tf.while_loop syntax; it's just an illustration.

Using eager execution would not help here, since the target is to run the optimizer in an tf.while_loop inside the computational graph. 
The eager execution would not stop me from using Python evaluations; instead it enforces the usage of Python loops.

I'm not sure, if we are talking about the same problem...

**Comment by mrry** [comment]
Unfortunately what you are asking for would be a large change to the semantics of graph-mode execution and `tf.while_loop()`: currently any tensors that precede the `tf.while_loop()` and are captured inside the body of the loop are captured by value and treated as a **constant** for the rest of the loop. Therefore both the `iterator.get_next()` **and** the calculation of `loss` must also be moved inside the `tf.while_loop()` body in order to process different examples in each iteration. 

I'm not aware of any plans to change this behavior.

/cc @skye, who maintains the control flow operators, and can correct me if I'm wrong here.

**Comment by skye** [comment]
Derek's correct, tensors defined outside a while loop cannot be (re)evaluated inside the loop, so any computation that happens inside the loop must be defined inside the loop as well. If you're feeling adventurous, you could try defining a TensorFlow function (https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/framework/function.py#L43) that performs the loss calculation in your example, and then call the function inside the while loop (and pass it the result of iterator.get_next()). Note that the Defun decorator is not part of the official API at this time though.

**Comment by skye** [comment]
Although I'm not sure using a TensorFlow function would actually buy you anything here, you may as well use a regular Python function :)

**Comment by Hoeze** [comment]
@mrry Thanks for the clarification, this explains my misunderstandings at the beginning :)
@skye That's an interesting feature, I'll definitely try that. You really have to promote such features more :D
